Authorization _ этап 2

KudryashovAlexander opened the pull request:
Всем привет!

Изменения:
- удалил LaunchViewController, так как дублировался с SplashViewController
- добавил алерты при неправильном вводе логина и пароля во время входа/ регистрации/ при попытке запроса кода восстановления
- добавил 2 скрина для восстановления пароля, пока без сетевого взаимодействия (еще не готово на стороне бека)

KudryashovAlexander committed:
fix localizationString authorization

KudryashovAlexander committed:
Убран LaunchViewController

KudryashovAlexander committed:
add RecoveryStartScreen

KudryashovAlexander committed:
add RecoveryEndScreen

KudryashovAlexander committed:
Добавлены алерты:
- при неправильном вводе логина и пароля;
- при неуказанном email в режиме восстановления пароля
- при некорректном указании логина и пароля во время регистрации

KudryashovAlexander committed:
fix to swiftLint

KudryashovAlexander committed:
Merge branch 'dev' into authorization

# Conflicts:
#	DiscountsAndPromotionsApp/Navigation/AuthCoordinator.swift
#	DiscountsAndPromotionsApp/UI/Auth/AuthScreen/View/LoginViewController.swift

yakushef commented on DiscountsAndPromotionsApp/UI/Auth/RecoveryStart/View/RecoveryStartViewController.swift:
```diff
@@ -0,0 +1,341 @@
+
+    private lazy var textFeildsHStack: UIStackView = {
+        var textFields = [UITextField]()
+        for dataTextField in viewModel.dataTextFields {
```
Мне понравилось твое решение с DataTextField, но UI пока еще не работает так, как от него ждешь - можно в любой момент ввести в любое поле сколько угодно символов. Обычно в таких ситуациях 4 поля для цифр ведут себя как одно. Мне кажется, можно использовать комбайн-паблишеры текстовых полей чтобы определять когда какое становится first responder. Но мотет проще будет переопределить методы делегата, например textFieldDidChange.
![Simulator Screenshot - iPhone 15 - 2024-02-02 at 09 16 05](https://github.com/Mobile-app-promotions-and-discounts/IOSApp/assets/61228372/e23491cf-3bd4-49d9-ac47-3a343b4abcb2)

yakushef reviewed (changes requested):
Привет! Посмотрел код. Наконец-то нам понадобилось комбинировать паблишеры в комбайне) Все по-моему хорошо, кроме поведения текстовых полей при вводе кода восстановления, но мне кажется это должно быть можно решить дополнительной несложной проверкой. Из не относящегося в функционалу - я обратил внимание, что при переходе между страницами флоу авторизации/регистрации/восстановления на время анимации пропадают скругления краев у фона (точнее, под ними как будто появляется непрозрачный фон).

KudryashovAlexander committed:
- исправлено скругление вью при переходах между экранами
- добавлен родительский UIViewController для экранов авторизации и регистрации
- немного оптимизирован код

KudryashovAlexander committed:
Поправлена работа UITextField в RecoveryStartViewController
- первый текстфилд автоматически становиться первым ответчиком
- при нажатии одного символа автоматически переходит на следующий текстфилд
- после внесения символа в последний текстфилд клавиатура скрывается
- при нажатии на текстфилд текст удаляется

KudryashovAlexander commented:
Привет!
Поправил, теперь должно работать как надо.

1) Скругление добавил путем скругления view самого UINavigationController. После добавления почему-то поехали title у вьюшек. Добавил родительский вью и добавил расширение для UILabel, где можно задавать межстрочный интервал. Это немного убрало дублирование кода и привели вьюшки к единому формату.

2) Логика работы с UITextFields для кода восстановления:
- первый текстфилд автоматически становиться первым ответчиком
- при нажатии одного символа автоматически переходит на следующий текстфилд
- после внесения символа в последний текстфилд клавиатура скрывается
- при нажатии на текстфилд текст удаляется

Комбинацию паблишеров в комбайне добавил в прошлом ПР, там проверялось на внесение логина и пароля, от этого зависела возможность нажатия на кнопку.
Для текстфилдов кода восстановления не совсем получилось красиво, хотел найти более гибкое решение, которое подстраивалось бы под кол-во символов. Пока не обнаружил фукционал, где можно было бы отслеживать массив паблишеров.

yakushef reviewed (changes requested):
Привет! Со скруглениями теперь все выглядит супер презентабельно.

Но я побродил по экранам и выяснил, что если сначала открыть экран восстановления пароля, а затем вернуться и залогиниться, то после нажатия на кнопку входа ничего не происходит, так и висит экран авторизации. После того как закрываю и открываю приложение заново то я уже залогинен и все ок.

Воспроизвести краш при создании нового пользователя, о котором писали ребята я не смог, но после успешной регистрации экран выбора города закрывается или сам спустя секунда или по нажатии на любую кнопку. Я немного потерялся в объемах выполненных и оставшихся задач - экран с городом еще в работе? Я помню, что Макс в рамках работы над профилем занимался выбором города при редактировании профиля - можно попробовать использовать его наработки.

Поля ввода кода стали работать логично при вводе, но мне показалось странным поведение при редактировании. После корректировки символа следующий стирается, а кнопка backspace на клавиатуре не работает. При этом можно просто нажатием сделать "дырку" в одном из полей а потом уйти на другое поле... У меня лично нет уверенности в том, как по уму устроить работу этих полей чтобы это было максимально удобно. Думаю, стоит уточнить у ребят их UX, по какой точно логике должен работать ввод этого кода и корректировка. А еще через физическую клавиатуру я смог ввести туда буквы. Я не уверен, насколько такая ситуация реальна с настоящим устройством (например, через автозаполнение) и насколько есть резон делать еще проверку на то чтобы вводились только цифры, но стоит иметь это в виду. Отдельный вопрос, получится ли через автозаполнение вставить больше одного символа в поле? Возможно можно его просто отключить для этих полей от греха подальше)

Ну и плейсхолдеры я бы все же убрал - потому что в дизайне их нет, а появились они как временное решение пока была невозможна регистрация - чтобы не возвращаться к ним когда остальные задачи по авторизации будут закрыты. Но сделать это, конечно, можно без проблем и позже.

KudryashovAlexander committed:
1) Поправлены binding у LoginViewController и его вьюМодель.
Исправлена ошибка при которой если зайти на экран восстановления и перейти обратно при вводе логина и пароля экран не переходил на MainScreen.

2) Поправлена логика работы textField на экране восстановления:
Перенос на следующий textField происходит при условии что он пустой.

3) Удалены плейсхолдеры на экранах ввода/регистрации пользователя.

KudryashovAlexander commented:
Привет!

1) Баг при переходе с экрана восстановления обратно в экран ввода поправил.

2) Воспроизвести краш мне тоже не получилось. При успешной регистрации сейчас пользователь на 1 секунду видит экран успешной регистрации и после появляется главный экран. Геопозицию планирую подключить на следующем спринте. Посмотрел реализацию на экране профиля. Как понял, геопозиция сама определяет город и записывает его UserDefaults  и меняется вью. На мой взгляд, логика должна быть реализована немного по-другому. Вынести работу с геопозицией в отдельную сущность, получить геолокацию с устройства и передать его на сервер. Исходя из ответа сервера указать город в профиле и получить выборку ближайших магазинов на главном экране.

2) Поправлена логика работы textField на экране восстановления:
Перенос на следующий textField происходит при условии что он пустой. На мой взгляд это оптимально для пользователя. По поводу ввода букв, это очень частный случай, на симуляторе открывается клавиатура с числами. Могу добавить проверку на введение букв дополнительно, но кажется излишним, так как если ввести некорректно, то от сервера должна прийти с ошибка ввода, туда и планирую добавить алерт.

3) Удалены плейсхолдеры на экранах ввода/регистрации пользователя.

KudryashovAlexander committed:
Немного скорректировал названия функций в Login и Registration

yakushef reviewed (approved):
Теперь все круто!
